File: py/tests/cross_validation/python/dump_adjoint.py
#!/usr/bin/env python3
"""Python-side raw-value dump matching cpp/dump_adjoint.cc, for
cross-validating py.vector_operations.FluxToYVelocity / YVelocityToFlux
against the C++ reference (verifies YVelocityToFlux is the adjoint of
FluxToYVelocity, per basis vector, across a 3-level multigrid).

Run from the repo root:  python3 -m py.tests.cross_validation.python.dump_adjoint
"""

from py.grid import Grid
from py.flux import Flux
from py.scalar import Scalar
from py.direction import Direction
from py.vector_operations import FluxToYVelocity, YVelocityToFlux
from .inner_products import InnerProductFlux, InnerProductScalar


def AdjYVelToFlux(q):
    grid = q.getGrid()
    x = Scalar(grid)
    x.assign(0.0)
    for lev in range(x.Ngrid()):
        for i in range(1, x.Nx()):
            for j in range(1, x.Ny()):
                e = Scalar(grid)
                e.assign(0.0)
                e._data[lev][i - 1, j - 1] = 1.0
                p = Flux(grid)
                p.assign(0.0)
                YVelocityToFlux(e, p)
                a = InnerProductFlux(q, p)
                normsq = InnerProductScalar(e, e)
                x._data[lev][i - 1, j - 1] = a / normsq
    return x


def dumpFluxY(name, q):
    print(name)
    for lev in range(q.Ngrid()):
        for i in range(0, q.Nx()):
            for j in range(0, q.Ny() + 1):
                print(repr(q(lev, Direction.Y, i, j)))


def main():
    nx, ny, ngrid = 8, 8, 3
    length = 0.1
    grid = Grid(nx, ny, ngrid, length, 0.0, 0.0)

    mag = Flux(grid)
    err2 = Flux(grid)
    mag.assign(0.0)
    err2.assign(0.0)

    for lev in range(ngrid):
        for ind in range(err2.begin(Direction.Y), err2.end(Direction.Y)):
            e = Flux(grid)
            e.assign(0.0)
            e.set(lev, ind, value=1.0)
            x0 = Scalar(grid)
            x0.assign(0.0)
            FluxToYVelocity(e, x0)
            x = AdjYVelToFlux(e)
            mag.set(lev, ind, value=InnerProductScalar(x0, x0))

            diff = Scalar(grid)
            diff.assign(x)
            diff._data[...] = x._data - x0._data  # full multi-level difference, matching C++ x-x0
            err2.set(lev, ind, value=InnerProductScalar(diff, diff))

    dumpFluxY("DUMP_mag", mag)
    dumpFluxY("DUMP_err2", err2)


if __name__ == "__main__":
    main()

File: py_static/tests/cross_validation/cpp/dump_adjoint.cc
// dump_adjoint.cc
//
// Raw-value dump driver for cross-validating py/vector_operations.py's
// FluxToYVelocity / YVelocityToFlux against the C++ reference. Reproduces
// the adjointness check from test/CheckAdjoint.cc (verify that
// YVelocityToFlux is the adjoint of FluxToYVelocity, computed brute-force
// per basis vector, across a 3-level multigrid), but prints "mag" and
// "err2" as one full-precision value per line instead of via the original's
// printY() (default 6-significant-digit cout formatting).
//
// Build (from repo root):
//   g++ -Wall -O2 -I src -c py/tests/cross_validation/cpp/dump_adjoint.cc -o /tmp/dump_adjoint.o
//   g++ /tmp/dump_adjoint.o -L build -libpm -lfftw3 -lm -o /tmp/dump_adjoint
// Run:
//   /tmp/dump_adjoint > /tmp/dump_adjoint.out

#include <iostream>
#include <iomanip>
#include "ibpm.h"

using namespace std;
using namespace ibpm;

// Compute the adjoint of YVelocityToFlux, brute force:
// x(lev,i,j) = < q, p(lev,i,j) > / <e,e> where e is an orthogonal basis function
void AdjYVelToFlux( const Flux& q, Scalar& x ) {
    Scalar e( x.getGrid() );
    x = 0.;
    for (int lev=0; lev<x.Ngrid(); ++lev) {
        for (int i=1; i<x.Nx(); ++i) {
            for (int j=1; j<x.Ny(); ++j) {
                e = 0;
                e(lev,i,j) = 1.;
                Flux p(q.getGrid());
                p = 0;
                YVelocityToFlux( e, p );
                double a = InnerProduct( q, p );
                double normsq = InnerProduct( e, e );
                x(lev,i,j) = a / normsq;
            }
        }
    }
}

void dumpFluxY(const char* name, const Flux& q) {
    cout << name << endl;
    for (int lev=0; lev<q.Ngrid(); ++lev)
        for (int i=0; i<q.Nx(); ++i)
            for (int j=0; j<=q.Ny(); ++j)
                cout << setprecision(17) << q(lev,Y,i,j) << endl;
}

int main() {
    int nx=8, ny=8, ngrid=3;
    double length=0.1;
    Grid grid( nx, ny, ngrid, length, 0., 0. );
    Flux mag(grid);
    Flux err2(grid);
    mag = 0.;
    err2 = 0.;

    // Loop over all fluxes in the Y direction
    for (int lev=0; lev<ngrid; ++lev) {
        for (int ind=err2.begin(Y); ind < err2.end(Y); ++ind) {
            Flux e(grid);
            e = 0;
            e(lev,ind) = 1;
            Scalar x(grid);
            Scalar x0(grid);
            x = 0;
            x0 = 0;
            FluxToYVelocity( e, x0 );
            AdjYVelToFlux( e, x );
            mag(lev,ind) = InnerProduct(x0,x0);
            err2(lev,ind) = InnerProduct(x-x0,x-x0);
        }
    }
    dumpFluxY("DUMP_mag", mag);
    dumpFluxY("DUMP_err2", err2);
    return 0;
}
